src/nomad_measurements/utils.py

```diff
@@ -153,3 +160,304 @@ def get_bounding_range_2d(ax1, ax2):
+        valid_dataset_paths: list = None,
+        nexus: bool = False,
+    ):
+        """
+        Initialize the handler.
+
+        Args:
+            filename (str): The name of the auxiliary file.
+            archive (EntryArchive): The NOMAD archive.
+            logger (BoundLogger): A structlog logger.
+            valid_dataset_paths (list): The list of valid dataset paths.
+            nexus (bool): If True, the file is created as a NeXus file.
+        """
+        if not filename.endswith(('.nxs', '.h5')):
+            raise ValueError('Only .h5 or .nxs files are supported.')
+
+        self.data_file = filename
+        self.archive = archive
+        self.logger = logger
+        self.valid_dataset_paths = []
+        if valid_dataset_paths:
+            self.valid_dataset_paths = valid_dataset_paths
+        self.nexus = nexus
+
+        self.hdf5_datasets = collections.OrderedDict()
+        self.hdf5_attributes = collections.OrderedDict()
+
+    def add_dataset(  # noqa: PLR0913
+        self,
+        path: str,
+        data: Any,
+        archive_path: str = None,
+        internal_reference: bool = False,
+        validate_path: bool = True,
+        lazy: bool = True,
+    ):
+        """
+        Add a dataset to the HDF5 file. The dataset is written lazily (default) when
+        either `read_dataset` or `write_file` method is called. The `path` is validated
+        against the `valid_dataset_paths` if provided before adding the data.
+
+        Args:
+            path (str): The dataset path to be used in the HDF5 file.
+            data (Any): The data to be stored in the HDF5 file.
+            archive_path (str): The path of the quantity in the archive.
+            internal_reference (bool): If True, an internal reference is set to an
+                existing HDF5 dataset.
+            validate_path (bool): If True, the path is validated against the
+                `valid_dataset_paths`.
+            lazy (bool): If True, the file is not written immediately.
+        """
+        if not path:
+            self.logger.warning('HDF5 `path` must be provided.')
+            return
+
+        if validate_path and self.valid_dataset_paths:
+            if path not in self.valid_dataset_paths:
+                self.logger.warning(f'Invalid dataset path "{path}".')
+                return
+
+        dataset = dict(
+            data=data,
+            attrs={},
+            hdf5_path=(
+                f'/uploads/{self.archive.m_context.upload_id}/raw'
+                f'/{self.data_file}#{path}'
+            ),
+            archive_path=archive_path,
+            internal_reference=internal_reference,
+        )
+        # handle the pint.Quantity and add data
+        if isinstance(data, pint.Quantity):
+            dataset['data'] = data.magnitude
+            dataset['attrs'].update({'units': str(data.units)})
+
+        self.hdf5_datasets[path] = dataset
+
+        if not lazy:
+            self.write_file()
```

Comment:
if not lazy it writes the dataset write away but this should be popped right so it does not get written again?

Reply:
The logic for clearing out `hdf5_dataset` is part of the write functions. Once all the datasets are written into the file, the `hdf5_dataset` and `hdf5_attributes` dicts are set to empty dicts.

Reply:
ah i did not see this, i expected pop.